src/shell/shell.cpp

```diff
@@ -452,6 +453,49 @@ void DOS_Shell::SyntaxError()
 	WriteOut(MSG_Get("SHELL_SYNTAX_ERROR"));
 }
 
+static std_fs::path get_cmd_history_path()
+{
+	return get_platform_config_dir() / "cmd_history.txt";
+}
+
+void DOS_Shell::ReadCommandHistory()
+{
+	std::ifstream history_file(get_cmd_history_path());
+	if (history_file) {
+		std::string line;
+		while (getline(history_file, line)) {
+			if (!line.empty()) {
+				history.push_back(line);
+			}
+		}
+	}
+}
+
+void DOS_Shell::WriteCommandHistory()
+{
+	std_fs::path history_path = get_cmd_history_path();
+	std::ofstream history_file(history_path);
+	if (!history_file) {
+		LOG_WARNING("SHELL: Unable to write command history to file: %s", history_path.string().c_str());
+		return;
+	}
+	std::vector<std::string> trimmed_history;
+	trimmed_history.reserve(history.size());
+	for (std::string& str : history) {
+		if (!str.empty()) {
+			trimmed_history.push_back(str);
+		}
+	}
+	size_t start = 0;
+	size_t size = trimmed_history.size();
+	if (size > CMD_MAX_HISTORY_FILE_LINES) {
+		start = size - CMD_MAX_HISTORY_FILE_LINES;
+	}
+	for (size_t i = start; i < size; i++) {
+		history_file << trimmed_history[i] << std::endl;
+	}
```

Comment:
Not sure about the usage of C++ streams. @kcgen , @dreamer ?

Reply:
I saw existing code in `src/misc/setup.cpp` was using C++ streams to parse the config file so that along with the fact that this is already using `std::string`, I thought C++ streams would be the simplest solution.

The output would be simple enough to switch to C-style file I/O.  Input is a little trickier since there's not a good "getline" function in C.  "getline" itself is POSIX only so it would need to use `fgets` alongside a temporary `char[] buffer` or something.

Just let me know what you prefer.

Reply:
@kcgen @dreamer ? ☝🏻 I'm 50/50.

Reply:
Right; we're not gaining any benefit from having streaming operators processing disparate types (the 'if/ofstream implementation come with a lot of bloat where they are used).

The history is just a list of strings, so hopefully can avoid using streams.

Reply:
I wonder if the copy of the history (trimmed_history) can be avoided - I /think/ it programmatically guaranteed that no empty lines exist as part of the prompt parsing.

However, we can't trust the incoming file (users might be tampering with it or it might be corrupt), so when reading, we should strip off empty lines and same with line exceeding our prompt length (I'm not sure of that.. might have to poke into the shell code).

For the writing, portion (after the check-and-returns), maybe this would :

```c++
constexpr size_t MaxHistoryFileLines = 500;
const auto num_lines = std::min(history.size(), MaxHistoryFileLines);

// the first line we'll write, iterated in the loop
const auto line = strings.end() - num_lines;

while (line != strings.end()) {
    history_file.write(line->c_str(), line->length());
    history_file.write("\n", 1);
    ++line;
}

history_file.close();
```

(I'm not sure if the close() is needed.. perhaps that's done on scope closure anyway)

Reply:
I only added `trimmed_history` in there after I tested and saw blank lines in the file.  It looks like items get added to the history before any parsing is done.

Reply:
Good catch, @weirddan455. I now understand the need for the trimmed history :-) (sorry!)

I'm thinking we only need the _indexes_ of the populated lines.

Here's one approach that can be dropped in after the ofstream check. 

```c++
auto get_populated_indices = [](const auto& lines, const auto max_indices) {
	std::stack<size_t> indices = {};
	auto reverse_i = lines.size();
	while (reverse_i-- != 0 && indices.size() < max_indices) {
		if (!lines[reverse_i].empty()) {
			indices.push(reverse_i);
		}
	}
	return indices;
};

constexpr size_t MaxHistoryFileLines = 500;
auto indices = get_populated_indices(history, MaxHistoryFileLines);

while (!indices.empty()) {
	const auto i = indices.top();
	indices.pop();

	history_file.write(history[i].c_str(), history[i].length());
	history_file.write("\n", 1);
}
```

It needs `#include <stack>` (for last-in-first-out behavior).

Reply:
I mean, yeah that works.  I'm thinking the strings will be small enough that it probably won't matter much in practice.  If you have mostly strings less than 8 characters, storing a `size_t` would actually be more memory.

I think I could actually re-write it to avoid a 2nd data structure at all if you think that's something worth doing.  It would just need 2 iterations checking `str.empty()` through `history`.  One to find the starting index and then one that actually writes it out to the file.

Reply:
Yup; good point @weirddan455 - let's go w/ your existing implementation. Nice and clear, and zero concern about speed.

Reply:
Although, suggest using the `.write()` calls, if you can.